py/nlr.c

```diff
@@ -50,6 +50,28 @@ void nlr_pop(void) {
     *top = (*top)->prev;
 }
 
+void nlr_push_jump_callback(nlr_jump_callback_t *node, nlr_jump_callback_fun_t fun) {
+    nlr_jump_callback_t **top = &MP_STATE_THREAD(nlr_jump_callback_top);
+    node->prev = *top;
+    node->fun = fun;
+    *top = node;
+}
+
+void nlr_pop_jump_callback(void) {
+    nlr_jump_callback_t **top = &MP_STATE_THREAD(nlr_jump_callback_top);
+    *top = (*top)->prev;
+}
+
+void nlr_call_jump_callbacks(nlr_buf_t *nlr) {
+    nlr_jump_callback_t **top = &MP_STATE_THREAD(nlr_jump_callback_top);
+    // Assumes descending stack.
```

Comment:
This comment is doing a lot of heavy lifting :) Even knowing what it had to be doing it took me a bit of thinking to see the chain.

It's a bit subtle what's going on here, but my understanding that `&ctx.callback` passed to `nlr_push_jump_callback` is a pointer to a member of a stack-allocated struct. And this must be deeper on the stack than the enclosing nlr buffer, so that tells it when to stop walking.

Reply:
Yes indeed, it could definitely do with some comments.  Also the requirement that the pointer passed to `nlr_push_jump_callback()` must be C stack allocated.

I also had the idea that `nlr_buf_t` itself could store the head of the linked list of callbacks to call, which then acts as a natural barrier for which callbacks to run when that NLR is caught.  But that adds a word to all `nlr_buf_t` uses.  Whereas the solution I went with here is a single pointer (albeit possibly less portable).

Reply:
I've added comments.